In [5]:
import numpy as np
from scipy.optimize import linprog

# https://docs.scipy.org/doc/scipy/tutorial/optimize.html#linear-programming-linprog

# metoda linprog rozwiązuje zadanie ZPL w postaci:
# min funkcję celu c
# przy ograniczeniach:
# nierównosciowych   A_ub <= b_ub
# równościowych A_eq = b_eq
# bounds ograniczenia na zmienne decyzyjne l <= x <= u -- domyślnie x >= 0
# method "revised simplex", "simplex", "interior-point"
# options  opcje dodatkowe np. czy rozwiazanie ma byc całkowitoliczbowe, boolowskie itp.

In [6]:
f_celu = [30, 32, 45]
lewa_strona_nier = [[ 20,  12, 7],  [0.8, 1.3, 0.25],   [-1.3, -1.1, -4], [-0.2, -1.1, -0.5]]
liczby_nier = [12,   1,   -2,  -0.5]
lewa_strona_rown = [[1, 1, 1]]
liczby_rown = [1]
ogr_zmiennych = [(0, float("inf")),  (0, float("inf")), (0, float("inf"))] # zmienne wieksze równe 0, można pominąć jesli nie ma dodatkowych ograniczeń

opt = linprog(c=f_celu, A_ub=lewa_strona_nier , b_ub=liczby_nier,
           A_eq=lewa_strona_rown, b_eq=liczby_rown, bounds=ogr_zmiennych,
            method="simplex")

print(opt)

     con: array([1.11022302e-16])
     fun: 35.49586776859504
 message: 'Optimization terminated successfully.'
     nit: 5
   slack: array([0.        , 0.1053719 , 0.        , 0.25413223])
  status: 0
 success: True
       x: array([0.18595041, 0.51652893, 0.29752066])


In [7]:
print(opt.fun)
print(opt.x)

35.49586776859504
[0.18595041 0.51652893 0.29752066]


In [8]:
# zmienne wieksze równe 0, można pominąć jesli nie ma dodatkowych ograniczeń

opt = linprog(c=f_celu, A_ub=lewa_strona_nier , b_ub=liczby_nier,
           A_eq=lewa_strona_rown, b_eq=liczby_rown, 
            method="simplex")

print(opt.fun)
print(opt.x)

35.49586776859504
[0.18595041 0.51652893 0.29752066]


# wyniki na wyjściu:
#     con: array([1.11022302e-16]) -- reszty ograniczeń równościowych: b_eq - A_eq @ x.
#     fun: 35.49586776859504 -- wartośc funkcji celu
# message: 'Optimization terminated successfully.' -- opis stanu algorytmu
#     nit: 5  -- liczba iteracji
#   slack: array([0.        , 0.1053719 , 0.        , 0.25413223]) -- reszty ograniczen nieównościowych  b_ub - A_ub @ x.
#  status: 0
# success: True
#       x: array([0.18595041, 0.51652893, 0.29752066])  -- wartości zmiennych decyzyjnych

In [9]:
#zagadnienie tarnsportowe

dost = 4
odb = 5
f_celu = [3, 2,	3,	2,	1,
4,	2,	4,	2,	3,
3,	1,	5,	4,	3,
2,	4,	1,	2,	2]
def zdefiniuj_lewa_strona_rownosci(dost, odb):
    line = [0 for i in range(dost*odb)]
    lewa_strona_rown = []
    for j in range(dost):
        line_rob = line.copy()
        for k in range(odb):
            line_rob[odb*j+k] = 1
        lewa_strona_rown.append(line_rob)
    for j in range(odb):
        line_rob = line.copy()
        for k in range(dost):
            line_rob[odb*k+j] = 1
        lewa_strona_rown.append(line_rob)
    return lewa_strona_rown
    
lewa_strona_rown = zdefiniuj_lewa_strona_rownosci(dost, odb)
liczby_rown = [20, 7, 15, 8, 8,	12,	10,	5,	15]
# ogr_zmiennych = [(0, float("inf")) for i in range(dost*odb)] # zmienne wieksze równe zero, przyjmowane jest domyślnie, więc można pominąć

opt = linprog(c=f_celu, 
           A_eq=lewa_strona_rown, b_eq = liczby_rown, method="simplex")

print('całkowity koszt przewozu', opt.fun)
print('dostawca1', opt.x[0:5])
print('dostawca2', opt.x[5:10])
print('dostawca3', opt.x[10:15])
print('dostawca4', opt.x[15:])


całkowity koszt przewozu 77.0
dostawca1 [ 5.  0.  0.  0. 15.]
dostawca2 [0. 0. 2. 5. 0.]
dostawca3 [ 3. 12.  0.  0.  0.]
dostawca4 [0. 0. 8. 0. 0.]


C:\Users\Bozena\AppData\Local\Temp\ipykernel_16820\2822140377.py:28: OptimizeWarning: A_eq does not appear to be of full row rank. To improve performance, check the problem formulation for redundant equality constraints.
  opt = linprog(c=f_celu,
